Feature/opencl (#30)
ULHPC/tutorials: docs/opencl/README.md

> [![By ULHPC](https://img.shields.io/badge/by-ULHPC-blue.svg)](https://hpc.uni.lu) [![Licence](https://img.shields.io/badge/license-GPL--3.0-blue.svg)](http://www.gnu.org/licenses/gpl-3.0.html) [![GitHub issues](https://img.shields.io/github/issues/ULHPC/tutorials.svg)](https://github.com/ULHPC/tutorials/issues/) [![](https://img.shields.io/badge/slides-PDF-red.svg)](https://github.com/ULHPC/tutorials/raw/devel/cuda/slides.pdf) [![Github](https://img.shields.io/badge/sources-github-green.svg)](https://github.com/ULHPC/tutorials/tree/devel/cuda/) [![Documentation Status](http://readthedocs.org/projects/ulhpc-tutorials/badge/?version=latest)](http://ulhpc-tutorials.readthedocs.io/en/latest/cuda/) [![GitHub forks](https://img.shields.io/github/stars/ULHPC/tutorials.svg?style=social&label=Star)](https://github.com/ULHPC/tutorials)

Reviewer: ```suggestion
[![By ULHPC](https://img.shields.io/badge/by-ULHPC-blue.svg)](https://hpc.uni.lu) [![Licence](https://img.shields.io/badge/license-GPL--3.0-blue.svg)](http://www.gnu.org/licenses/gpl-3.0.html) [![GitHub issues](https://img.shields.io/github/issues/ULHPC/tutorials.svg)](https://github.com/ULHPC/tutorials/issues/) [![](https://img.shields.io/badge/slides-PDF-red.svg)](https://github.com/ULHPC/tutorials/raw/devel/opencl/slides.pdf) [![Github](https://img.shields.io/badge/sources-github-green.svg)](https://github.com/ULHPC/tutorials/tree/devel/opencl/) [![Documentation Status](http://readthedocs.org/projects/ulhpc-tutorials/badge/?version=latest)](http://ulhpc-tutorials.readthedocs.io/en/latest/opencl/) [![GitHub forks](https://img.shields.io/github/stars/ULHPC/tutorials.svg?style=social&label=Star)](https://github.com/ULHPC/tutorials)
```

# Introduction to OpenCL Programming (C/C++)

     Copyright (c) T. Carneiro, L. Koutsantonis, 2013-2021 UL HPC Team <hpc-sysadmins@uni.lu>

> [![](./cover_slides.png)](./PS10b2.pdf)

Reviewer: ```suggestion
[![](https://github.com/ULHPC/tutorials/raw/devel/opencl/cover_slides.png)](https://github.com/ULHPC/tutorials/raw/devel/opencl/PS10b2.pdf)
```

[0]: ./code "code"
[1]: https://programmerclick.com/article/47811146604/ "Tutorial: Simple start with OpenCL and C++" 
[2]: https://www.khronos.org/registry/OpenCL/ "Khronos OpenCL Working Group. The OpenCL Specification (Oct. 2021)"
[3]: https://www.eriksmistad.no/getting-started-with-opencl-and-gpu-computing/ "Smistad, E. Getting started with OpenCL and GPU Computing, Feb. 22, 2018 (Access on Oct. 28, 2021)."
[4]: https://www.nersc.gov/assets/pubs_presos/MattsonTutorialSC14.pdf "Mattson, T., McIntosh-Smith, S.,  Koniges, A. OpenCL: a Hands-on Introduction"

- [Tutorial: Simple start with OpenCL and C++][1]
- [Khronos OpenCL Working Group. The OpenCL Specification (Oct. 2021)][2]
- [Smistad, E. Getting started with OpenCL and GPU Computing, Feb. 22, 2018 (Access on Oct. 28, 2021).][3]
- [Mattson, T., McIntosh-Smith, S.,  Koniges, A. OpenCL: a Hands-on Introduction][4]

--------------------
# Programming Pre-requisites: #

- C/C++ language and, due to the lower level of OpenCL, 
- (preferable) knowledge in other frameworks for heterogeneous programming, such as CUDA, OpenACC or SYCL. 

--------------------

# Connecting to a GPU-enable node on the Iris Cluster #

> Ensure you can [connect to the UL HPC clusters](https://hpc.uni.lu/users/docs/access.html).

Reviewer: ```suggestion
Ensure you can [connect to the UL HPC clusters](https://hpc-docs.uni.lu/connect/access/).

In particular, recall that the `module` command **is not** available on the access frontends.

```bash

> ### Access to ULHPC cluster - here iris
> (laptop)$>  ssh -p 8022 user@access-iris.uni.lu

Reviewer: ```suggestion
### Access to ULHPC cluster - here iris as it is the only supercomputer with GPU accelerators
(laptop)$>  ssh iris-cluster
```

```

Now you'll need to pull the latest changes in your working copy of the [ULHPC/tutorials](https://github.com/ULHPC/tutorials) you should have cloned in `~/git/github.com/ULHPC/tutorials` (see ["preliminaries" tutorial](../../preliminaries/))

``` bash
(access)$> cd ~/git/github.com/ULHPC/tutorials
(access)$> git pull
```

## Accessing a GPU-equipped node of the Iris cluster

To reserve a node with one GPU for interactive development, load the necessary modules, and save them for a quick restore.

As usual, more information can be found in the [documentation][5].

> [5]: https://hpc.uni.lu/users/docs/gpu.html

Reviewer: ```suggestion
[5]: https://hpc-docs.uni.lu/jobs/gpu/
```
